File: bindings/tauri/tauri-plugin-kevy/src/reply.rs
//! Decode the raw RESP bytes that [`Store::dispatch_argv`] produces into a
//! JSON-friendly reply the webview can consume.
//!
//! The embedded `cmd` path returns the same wire bytes a kevy server would, so
//! we reuse the reference RESP parser (`kevy_resp::parse_reply`) — no bespoke
//! decoder — and then re-shape its [`Reply`] into a serde enum. Byte strings
//! (`Bulk`, `Simple`, …) serialise as JSON number arrays so they stay
//! **binary-safe** across the IPC boundary; the guest binding converts them
//! back to `Uint8Array` (see `guest-js/src/reply.ts`).
//!
//! `dispatch_argv` speaks RESP2 (the embedded store never negotiates `HELLO 3`),
//! so in practice only `Simple`/`Error`/`Int`/`Bulk`/`Nil`/`Array` appear; the
//! remaining RESP3 variants are mapped for completeness.

use kevy_resp::{parse_reply, Reply};
use serde::Serialize;

use crate::error::{Error, Result};

/// A RESP reply re-shaped for JSON transport to the webview.
///
/// Serialises as `{ "type": "<tag>", … }`, mirroring `docs/client-contract.md`
/// §4.1. All byte payloads are JSON arrays of `u8` (binary-safe).
#[derive(Debug, Clone, Serialize)]
#[serde(tag = "type", rename_all = "lowercase")]
pub enum JsonReply {
    /// `+OK` — a simple status string.
    Simple {
        /// Status bytes (without the leading `+`).
        bytes: Vec<u8>,
    },
    /// `-ERR …` — an error frame (successful round-trip carrying an error).
    Error {
        /// Error text bytes (without the leading `-`).
        bytes: Vec<u8>,
    },
    /// `:N` — a 64-bit integer.
    Int {
        /// The integer value.
        int: i64,
    },
    /// `$len …` — a bulk (byte) string.
    Bulk {
        /// Payload bytes.
        bytes: Vec<u8>,
    },
    /// `$-1` / `*-1` — the RESP2 null sentinel.
    Nil,
    /// `*N …` — an array of replies.
    Array {
        /// Element replies.
        items: Vec<JsonReply>,
    },
    /// `%N …` (RESP3) — a map of reply pairs.
    Map {
        /// `[key, value]` reply pairs, order preserved.
        pairs: Vec<(JsonReply, JsonReply)>,
    },
    /// `~N …` (RESP3) — a set of replies.
    Set {
        /// Member replies.
        items: Vec<JsonReply>,
    },
    /// `,x` (RESP3) — a double.
    Double {
        /// The double value.
        double: f64,
    },
    /// `#t`/`#f` (RESP3) — a boolean.
    Boolean {
        /// The boolean value.
        boolean: bool,
    },
    /// `=…` (RESP3) — a verbatim string with a 3-char format tag.
    Verbatim {
        /// The 3-char format tag (e.g. `"txt"`, `"mkd"`).
        format: String,
        /// Payload bytes (after the `:` separator).
        bytes: Vec<u8>,
    },
    /// `(…` (RESP3) — an arbitrary-precision integer, carried as digit bytes.
    Bignumber {
        /// Raw digit bytes.
        bytes: Vec<u8>,
    },
    /// `_` (RESP3) — a true null.
    Null,
    /// `>N …` (RESP3) — an out-of-band push frame.
    Push {
        /// Element replies.
        items: Vec<JsonReply>,
    },
    /// `!…` (RESP3) — a length-prefixed blob error.
    Bloberror {
        /// Error bytes.
        bytes: Vec<u8>,
    },
}

impl From<Reply> for JsonReply {
    fn from(r: Reply) -> Self {
        match r {
            Reply::Simple(b) => JsonReply::Simple { bytes: b },
            Reply::Error(b) => JsonReply::Error { bytes: b },
            Reply::Int(n) => JsonReply::Int { int: n },
            Reply::Bulk(b) => JsonReply::Bulk { bytes: b },
            Reply::Nil => JsonReply::Nil,
            Reply::Array(v) => JsonReply::Array { items: map_all(v) },
            Reply::Map(p) => JsonReply::Map { pairs: map_pairs(p) },
            Reply::Set(v) => JsonReply::Set { items: map_all(v) },
            Reply::Double(d) => JsonReply::Double { double: d },
            Reply::Boolean(b) => JsonReply::Boolean { boolean: b },
            Reply::Verbatim { fmt, data } => JsonReply::Verbatim {
                format: String::from_utf8_lossy(&fmt).into_owned(),
                bytes: data,
            },
            Reply::BigNumber(b) => JsonReply::Bignumber { bytes: b },
            Reply::Null => JsonReply::Null,
            Reply::Push(v) => JsonReply::Push { items: map_all(v) },
            Reply::BlobError(b) => JsonReply::Bloberror { bytes: b },
        }
    }
}

fn map_all(v: Vec<Reply>) -> Vec<JsonReply> {
    v.into_iter().map(JsonReply::from).collect()
}

fn map_pairs(p: Vec<(Reply, Reply)>) -> Vec<(JsonReply, JsonReply)> {
    p.into_iter()
        .map(|(k, v)| (JsonReply::from(k), JsonReply::from(v)))
        .collect()
}

/// Parse the single RESP reply `dispatch_argv` wrote into `buf`.
///
/// A malformed or truncated frame is a `Protocol` error (it should never
/// happen — the embedded store emits well-formed RESP — but we surface it
/// honestly rather than panic).
pub fn decode(buf: &[u8]) -> Result<JsonReply> {
    match parse_reply(buf) {
        Ok(Some((reply, _used))) => Ok(reply.into()),
        Ok(None) => Err(Error::Protocol("truncated RESP reply from embedded store".into())),
        Err(e) => Err(Error::Protocol(format!("malformed RESP reply: {e:?}"))),
    }
}

#[cfg(test)]
#[path = "reply_tests.rs"]
mod tests;

File: bindings/tauri/guest-js/src/reply.ts
// The decoded RESP reply the plugin's `cmd` command returns.
//
// Mirrors the Rust `JsonReply` (see tauri-plugin-kevy/src/reply.rs) and the
// `Reply` taxonomy in docs/client-contract.md §4.1. Byte payloads arrive as
// JSON number arrays for binary safety; the helpers below turn them back into
// `Uint8Array` / `string`.

/** A byte payload as it crosses the IPC boundary: a plain array of 0–255. */
export type ByteArray = number[]

/** The decoded reply, tagged by `type` (lowercase, matching the wire). */
export type Reply =
  | { type: 'simple'; bytes: ByteArray }
  | { type: 'error'; bytes: ByteArray }
  | { type: 'int'; int: number }
  | { type: 'bulk'; bytes: ByteArray }
  | { type: 'nil' }
  | { type: 'array'; items: Reply[] }
  | { type: 'map'; pairs: [Reply, Reply][] }
  | { type: 'set'; items: Reply[] }
  | { type: 'double'; double: number }
  | { type: 'boolean'; boolean: boolean }
  | { type: 'verbatim'; format: string; bytes: ByteArray }
  | { type: 'bignumber'; bytes: ByteArray }
  | { type: 'null' }
  | { type: 'push'; items: Reply[] }
  | { type: 'bloberror'; bytes: ByteArray }

const dec = new TextDecoder()
const enc = new TextEncoder()

/** Coerce a key/value argument (`string` | bytes) into the wire byte array. */
export function toByteArray(v: string | Uint8Array | ArrayBuffer | number[]): ByteArray {
  if (typeof v === 'string') return Array.from(enc.encode(v))
  if (v instanceof Uint8Array) return Array.from(v)
  if (v instanceof ArrayBuffer) return Array.from(new Uint8Array(v))
  return v
}

/** A reply's byte payload as a `Uint8Array`, or `null` for nil/null. */
export function replyBytes(r: Reply): Uint8Array | null {
  switch (r.type) {
    case 'bulk':
    case 'simple':
    case 'error':
    case 'verbatim':
    case 'bignumber':
    case 'bloberror':
      return Uint8Array.from(r.bytes)
    case 'nil':
    case 'null':
      return null
    default:
      throw new TypeError(`reply is ${r.type}, not a byte payload`)
  }
}

/** A reply's byte payload as a UTF-8 string, or `null` for nil/null. */
export function replyString(r: Reply): string | null {
  const b = replyBytes(r)
  return b === null ? null : dec.decode(b)
}

/** A reply's integer value. Throws if the reply is not an `int`. */
export function replyInt(r: Reply): number {
  if (r.type === 'int') return r.int
  throw new TypeError(`reply is ${r.type}, not an int`)
}

/** A reply's array items. Throws if the reply is not an `array`/`set`/`push`. */
export function replyArray(r: Reply): Reply[] {
  if (r.type === 'array' || r.type === 'set' || r.type === 'push') return r.items
  throw new TypeError(`reply is ${r.type}, not an array`)
}

/** Map an array reply to `(Uint8Array | null)[]` — the common bulk-list shape. */
export function replyBytesList(r: Reply): (Uint8Array | null)[] {
  return replyArray(r).map(replyBytes)
}

/** Map an array reply to `string[]` (dropping nils). */
export function replyStringList(r: Reply): string[] {
  return replyArray(r)
    .map(replyString)
    .filter((s): s is string => s !== null)
}
